crates/adapters/src/synchronizer/mod.rs

```diff
@@ -0,0 +1,180 @@
+mod error;
+mod rclone;
+
+use anyhow::Result;
+use dbsp::circuit::checkpointer::{CheckpointMetadata, CHECKPOINT_FILE_NAME};
+use error::SyncError;
+use feldera_adapterlib::errors::journal::ControllerError;
+use feldera_types::config::SyncConfig;
+use rclone::{RcloneLocation, RcloneSync};
+use serde::Deserialize;
+use std::{
+    ffi::OsStr,
+    fs::File,
+    io::BufReader,
+    path::{Path, PathBuf},
+    process::ExitStatus,
+};
+use tokio::io::AsyncReadExt;
+
+use crate::controller::STATE_FILE;
+
+#[derive(Debug, Deserialize)]
+struct PSpineFiles {
+    files: Vec<String>,
+}
+
+struct AllCheckpointFiles {
+    metadata: Vec<String>,
+    reference: Vec<String>,
+}
+
+impl AllCheckpointFiles {
+    fn from_checkpoint(chk: &uuid::Uuid, storage_path: &Path) -> Result<Self, SyncError> {
```

Comment:
This function seems to duplicate a lot of the code in `gather_batches_for_checkpoint` in `crates/dbsp/src/circuit/checkpointer.rs`. Ideally, we wouldn't want to duplicate code like that, so maybe that function should be factored out into something that can be used in both places.

I think that the main difference is that that function uses a storage backend instead of the file system directly. But it should not be a big deal to use a storage backend here, I think.